service/migration_manager.cc

```diff
@@ -1145,27 +1145,38 @@ future<schema_ptr> migration_manager::get_schema_for_write(table_schema_version
     }
 
     auto s = local_schema_registry().get_or_null(v);
-
-    if (s && s->is_synced()) {
-        co_return s;
-    }
-
     // `_enable_schema_pulls` may change concurrently with this function (but only from `true` to `false`).
     bool use_raft = !_enable_schema_pulls;
-
-    if (use_raft) {
-        // Schema is synchronized through Raft, so perform a group 0 read barrier.
-        // Batch the barriers so we don't invoke them redundantly.
+    if ((!s || !s->is_synced()) && use_raft) {
+        // batch group0 raft barriers
         co_await (as ? _group0_barrier.trigger(*as) : _group0_barrier.trigger());
     }
-
-    s = co_await get_schema_definition(v, dst, ms, _storage_proxy);
-
-    if (use_raft) {
-        // If Raft is used the schema is synced already (through barrier above), mark it as such.
-        co_await s->registry_entry()->maybe_sync([] { return make_ready_future<>(); });
-    } else {
-        co_await maybe_sync(s, dst);
+    if (!s) {
+        s = co_await get_schema_definition(v, dst, ms, _storage_proxy);
+    }
+    if (!s->is_synced()) {
+        if (use_raft) {
+            // Mark entry as synced we have already aquired the barrier
+            co_await s->registry_entry()->maybe_sync([] { return make_ready_future<>(); });
+        } else {
+            co_await maybe_sync(s, dst);
+        }
+    }
+    // here s is guaranteed to be valid and synced
+    if (s->is_view() && !s->view_info()->base_info()) {
+        // The way to get here is if the view schema was deactivated
+        // and reactivated again, or if we loaded it from the schema
+        // history.
+        auto& db = _storage_proxy.local_db();
+        // This line might throw a no_such_column_family but
+        // that is fine, if the schema is synced, it means that if
+        // we failed to get the base table, we learned about the base
+        // table not existing (which means that the view also doesn't exist
+        // any more), which means that this schema is actually useless for either
+        // read or write so we better throw than return an incomplete useless
+        // schema
+        schema_ptr base_schema = db.find_schema(s->view_info()->base_id());
+        s->view_info()->set_base_info(s->view_info()->make_base_dependent_view_info(*base_schema));
```

Comment:
This is done after s is marked as synced, after deferring point, and can race with other requests which may hit in the schema registry before we set the base info, and still trigger the abort.

Reply:
@tgrabiec 
I think that what you are asking is exactly what this code is trying to fix,
The bug was that prior to this patch we assumed that a `synced schema` == `schema with base info`
This proved to be wrong because of deactivate->activate (which doesn't restore the base info)
It is worth mentioning the the write workflow (and recently also the read) always gets its schema through this workflow, so following the logic of this patch (in high level):
1. Retrieve a synced schema (regardless of raft), at that point the registry entry can't be deactivated.
2. Fix the view info - which will also pin the base table in registry (if exists)
3. Either Return this schema to the caller or that at some point we except out due to stale schema (missing base which means that the view is also stale)

Maybe I am missing something and having a hard time to figure this out.
Would you describe the scenario you had in mind, that can lead to the crash? (with this code in)

Reply:
I was worried that a concurrent `local_schema_registry().get_or_load()` which is not going through get_schema_for_write() could get a schema which doesn't yet have the view attached, if it happens before we attach it here.

One such place is here:

```
schema_ptr schema::get_reversed() const {
    return local_schema_registry().get_or_load(reversed(_raw._version), [this] (table_schema_version) {
        return frozen_schema(make_reversed());
    });
}
```

Is it vulnerable too?

Reply:
Yes, it is vulnerable.
However:
1. I couldn't find where in the code this is actually used (except for tests).
2. In order to fail one will need to reverse the schema and then attempt to make a `global_ptr` out of it.

You are right that we probably need a more accurate solution to the problem (probably amend schema registry in some way to handle "post freeze".
However, the specific flow I fixed was found by investigating real failures.

Reply:
> Yes, it is vulnerable. However:
> 
> 1. I couldn't find where in the code this is actually used (except for tests).

It's used in reverse-order reads.

> 2. In order to fail one will need to reverse the schema and then attempt to make a `global_ptr` out of it.

Or just try to access the view info, which will be missing in the reversed schema, right?